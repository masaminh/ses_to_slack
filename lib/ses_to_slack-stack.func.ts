import type * as lambda from 'aws-lambda';
import { S3Client, GetObjectCommand, PutObjectCommand } from '@aws-sdk/client-s3';
import { SQSClient, SendMessageBatchRequestEntry, SendMessageBatchCommand } from '@aws-sdk/client-sqs';
import { Readable } from 'stream';
import MessageFormatter from './lambda/message_formatter';

export const handler: lambda.SESHandler = async (event) => {
  const bucketName = process.env.BUCKET ?? '';
  if (bucketName === '') {
    throw new Error('Environment Variable [BUCKET] is not set.');
  }

  const cdnBucketName = process.env.CDN_BUCKET ?? '';
  if (cdnBucketName === '') {
    throw new Error('Environment Variable [CDN_BUCKET] is not set.');
  }

  const messages = await Promise.all(
    event.Records.map(async (record, index): Promise<SendMessageBatchRequestEntry> => {
      const { messageId } = record.ses.mail;
      const s3 = new S3Client({});
      const output = await s3.send(
        new GetObjectCommand({ Bucket: bucketName, Key: messageId }),
      );

      if (!output.Body) {
        throw new Error('Body is empty.');
      }

      const messageInfo = await MessageFormatter.format(output.Body as Readable);
      const imageInfos = messageInfo.attachments.filter((x) => x.contentType.startsWith('image/'));
      imageInfos.forEach((x) => {
        s3.send(
          new PutObjectCommand({ Bucket: cdnBucketName, Key: `${messageId}/${x.fileName}`, Body: x.content }),
        );
      });

      return {
        Id: `${index}`,
        MessageBody: JSON.stringify({
          webhookname: process.env.WEBHOOK_NAME,
          message: messageInfo.text,
          images: imageInfos.map(
            (x) => ({ imageUrl: `https://${process.env.CDN_DOMAIN}/${messageId}/${x.fileName}`, imageName: x.fileName }),
          ),
        }),
      };
    }),
  );

  const sqs = new SQSClient({ region: 'ap-northeast-1' });
  await sqs.send(
    new SendMessageBatchCommand({
      QueueUrl: process.env.QUEUE_URL ?? '',
      Entries: messages,
    }),
  );
};
