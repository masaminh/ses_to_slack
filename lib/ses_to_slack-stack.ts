import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as lambdaNodejs from 'aws-cdk-lib/aws-lambda-nodejs';
import * as ses from 'aws-cdk-lib/aws-ses';
import * as sesActions from 'aws-cdk-lib/aws-ses-actions';
import * as s3 from 'aws-cdk-lib/aws-s3';
import * as sqs from 'aws-cdk-lib/aws-sqs';
import * as cloudfront from 'aws-cdk-lib/aws-cloudfront';
import config from 'config';

export class SesToSlackStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const ruleSetName = config.get<string>('ruleSetName');
    const recipient = config.get<string>('recipient');
    const stage = config.get<string>('stage');
    const webhookName = config.get<string>('webhook');
    const queueUrl = config.get<string>('queue');
    const queueArn = config.get<string>('queuearn');

    const bucket = new s3.Bucket(this, 's3', {
      lifecycleRules: [
        {
          expiration: cdk.Duration.days(30),
          noncurrentVersionExpiration: cdk.Duration.days(7),
        },
      ],
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      encryption: s3.BucketEncryption.S3_MANAGED,
      enforceSSL: true,
      versioned: true,
    });
    const bucketCdn = new s3.Bucket(this, 's3Cdn', {
      lifecycleRules: [
        {
          expiration: cdk.Duration.days(90),
          noncurrentVersionExpiration: cdk.Duration.days(7),
        },
      ],
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      encryption: s3.BucketEncryption.S3_MANAGED,
      enforceSSL: true,
      versioned: true,
    });

    const oai = new cloudfront.OriginAccessIdentity(this, 'oai');

    const distribution = new cloudfront.CloudFrontWebDistribution(this, 'distribution', {
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: bucketCdn,
            originAccessIdentity: oai,
          },
          behaviors: [
            { isDefaultBehavior: true },
          ],
        },
      ],
    });

    const lambdaFunction = new lambdaNodejs.NodejsFunction(this, 'func', {
      timeout: cdk.Duration.seconds(30),
      environment: {
        BUCKET: bucket.bucketName,
        WEBHOOK_NAME: webhookName,
        QUEUE_URL: queueUrl,
        CDN_BUCKET: bucketCdn.bucketName,
        CDN_DOMAIN: distribution.distributionDomainName,
      },
      tracing: lambda.Tracing.ACTIVE,
      bundling: {
        forceDockerBundling: false,
      },
    });

    bucket.grantRead(lambdaFunction);
    bucketCdn.grantWrite(lambdaFunction);

    const queue = sqs.Queue.fromQueueArn(this, 'queue', queueArn);
    queue.grantSendMessages(lambdaFunction);

    const ruleSet = ses.ReceiptRuleSet.fromReceiptRuleSetName(this, 'ruleset', ruleSetName);
    ruleSet.addRule('rule', {
      receiptRuleName: `ToSlackRule${stage}`,
      recipients: [recipient],
      actions: [
        new sesActions.S3({
          bucket,
        }),
        new sesActions.Lambda({
          function: lambdaFunction,
        }),
      ],
    });
  }
}
