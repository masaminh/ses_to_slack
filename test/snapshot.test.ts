import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SesToSlackStack } from '../lib/ses_to_slack-stack';

test('snapshot test', () => {
  const app = new cdk.App();
  const stack = new SesToSlackStack(app, 'MyTestStack');
  // スタックからテンプレート(JSON)を生成
  const template = Template.fromStack(stack).toJSON();

  // 生成したテンプレートとスナップショットが同じか検証
  expect(template).toMatchSnapshot();
});
