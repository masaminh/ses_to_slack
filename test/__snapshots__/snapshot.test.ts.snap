// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "distribution114A0A2A": {
      "Properties": {
        "DistributionConfig": {
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "MyTestStackdistributionOrigin1C99B7CCA",
            "ViewerProtocolPolicy": "allow-all",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "s3Cdn5EFD3905",
                  "RegionalDomainName",
                ],
              },
              "Id": "MyTestStackdistributionOrigin1C99B7CCA",
              "S3OriginConfig": {
                "OriginAccessIdentity": {
                  "Fn::Join": [
                    "",
                    [
                      "origin-access-identity/cloudfront/",
                      {
                        "Ref": "oai7A77E975",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "PriceClass": "PriceClass_100",
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "funcAllowSesC2855407": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "funcC3A0C2E2",
            "Arn",
          ],
        },
        "Principal": "ses.amazonaws.com",
        "SourceAccount": {
          "Ref": "AWS::AccountId",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "funcC3A0C2E2": {
      "DependsOn": [
        "funcServiceRoleDefaultPolicyF8358F12",
        "funcServiceRoleA96CCB44",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "HASH-REPLACED.zip",
        },
        "Environment": {
          "Variables": {
            "BUCKET": {
              "Ref": "s3100BEDFB",
            },
            "CDN_BUCKET": {
              "Ref": "s3Cdn5EFD3905",
            },
            "CDN_DOMAIN": {
              "Fn::GetAtt": [
                "distribution114A0A2A",
                "DomainName",
              ],
            },
            "QUEUE_URL": "https://sqs.ap-northeast-1.amazonaws.com/123456789012/Queue",
            "WEBHOOK_NAME": "Webhook",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "funcServiceRoleA96CCB44",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 30,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "funcServiceRoleA96CCB44": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "funcServiceRoleDefaultPolicyF8358F12": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "s3100BEDFB",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "s3100BEDFB",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "s3Cdn5EFD3905",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "s3Cdn5EFD3905",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:sqs:ap-northeast-1:123456789012:Queue",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "funcServiceRoleDefaultPolicyF8358F12",
        "Roles": [
          {
            "Ref": "funcServiceRoleA96CCB44",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "oai7A77E975": {
      "Properties": {
        "CloudFrontOriginAccessIdentityConfig": {
          "Comment": "Allows CloudFront to reach the bucket",
        },
      },
      "Type": "AWS::CloudFront::CloudFrontOriginAccessIdentity",
    },
    "rulesetruleAFCB6A4F": {
      "DependsOn": [
        "funcAllowSesC2855407",
        "s3PolicyED93D1AD",
      ],
      "Properties": {
        "Rule": {
          "Actions": [
            {
              "S3Action": {
                "BucketName": {
                  "Ref": "s3100BEDFB",
                },
              },
            },
            {
              "LambdaAction": {
                "FunctionArn": {
                  "Fn::GetAtt": [
                    "funcC3A0C2E2",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "Enabled": true,
          "Name": "ToSlackRuleTest",
          "Recipients": [
            "test@example.com",
          ],
        },
        "RuleSetName": "default-rule-set",
      },
      "Type": "AWS::SES::ReceiptRule",
    },
    "s3100BEDFB": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": {
          "Rules": [
            {
              "ExpirationInDays": 30,
              "NoncurrentVersionExpiration": {
                "NoncurrentDays": 7,
              },
              "Status": "Enabled",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "s3Cdn5EFD3905": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "ServerSideEncryptionByDefault": {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": {
          "Rules": [
            {
              "ExpirationInDays": 90,
              "NoncurrentVersionExpiration": {
                "NoncurrentDays": 7,
              },
              "Status": "Enabled",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "s3CdnPolicy6430CEB8": {
      "Properties": {
        "Bucket": {
          "Ref": "s3Cdn5EFD3905",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "s3Cdn5EFD3905",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "s3Cdn5EFD3905",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": {
                "CanonicalUser": {
                  "Fn::GetAtt": [
                    "oai7A77E975",
                    "S3CanonicalUserId",
                  ],
                },
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "s3Cdn5EFD3905",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "s3PolicyED93D1AD": {
      "Properties": {
        "Bucket": {
          "Ref": "s3100BEDFB",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "s3100BEDFB",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "s3100BEDFB",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:PutObject",
              "Condition": {
                "StringEquals": {
                  "aws:Referer": {
                    "Ref": "AWS::AccountId",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "ses.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "s3100BEDFB",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
